src/services/ExperianService.ts

```diff
@@ -0,0 +1,124 @@
+    	this.logger = logger;
+		this.dynamo = dynamoDbClient;
+		this.experianBaseUrl = experianBaseUrl;
+		this.experianTokenTableName = experianTokenTableName;
+	}
+
+	static getInstance(logger: Logger, experianBaseUrl: string, dynamoDbClient: DynamoDBDocument, experianTokenTableName: string): ExperianService {
+    	if (!ExperianService.instance) {
+    		ExperianService.instance = new ExperianService(logger, experianBaseUrl, dynamoDbClient, experianTokenTableName);
+    	}
+    	return ExperianService.instance;
+	}
+
+	async generateExperianToken(clientUsername: string, clientPassword: string, clientId: string, clientSecret: string): Promise<ExperianTokenResponse> {
+    	this.logger.info({ message: `Checking ${this.experianTokenTableName} for valid token` });
+		const token = await this.checkExperianToken();
+		if (token) {
+			return token;
+		} else {		
+			this.logger.info("requestExperianToken - no valid token - trying to generate new Experian token");
+			try {
+				const params = {
+					username: clientUsername,
+					password: clientPassword,
+					client_id: clientId,
+					client_secret: clientSecret,
+				};
+
+				const config: AxiosRequestConfig<any> = {
+    				headers: {
+    					"Content-Type": "application/json",
+						"X-Correlation-Id": randomUUID(),
+						"X-User-Domain": "cabinetofficegds.com",
+    				},
+				};
+
+				const { data }: { data: ExperianTokenResponse } = await axios.post(
+					`${this.experianBaseUrl}${Constants.EXPERIAN_TOKEN_ENDPOINT_PATH}`,
+					params,
+					config,
+				);
+				this.logger.info("Received response from experian token endpoint");
+				await this.saveExperianToken(data);
+				return data;
+			} catch (error: any) {
+				this.logger.error({ message: "Error generating experian token", statusCode: error?.response?.status, messageCode: MessageCodes.FAILED_GENERATING_EXPERIAN_TOKEN });
+				throw new AppError(HttpCodesEnum.SERVER_ERROR, "Error generating experian token");
+			}
+		}
+	}
+
+	async saveExperianToken(token : ExperianTokenResponse): Promise<void> {
+		this.logger.info({ message: `Updating ${this.experianTokenTableName} with new Experian token` });
+
+		const putExperianTokenCommand = new PutCommand({
+			TableName: this.experianTokenTableName,
+			Item: { 
+				id: "1",
+				issued_at: token.issued_at,
+				expires_in: token.expires_in,
+				token_type: token.token_type,
+				access_token: token.access_token,
+				refresh_token: token.refresh_token,
+			},
+		});
+
+		try {
+			await this.dynamo.send(putExperianTokenCommand);
+			this.logger.info({ message: "Stored new Experian token" });
+		} catch (error) {
+			this.logger.error({ message: "Error storing new Experian token", messageCode: MessageCodes.FAILED_GENERATING_EXPERIAN_TOKEN, error });
+			throw new AppError(HttpCodesEnum.SERVER_ERROR, "Error storing new Experian token");
+		}
+	}
+
+	async checkExperianToken(): Promise<ExperianTokenResponse | undefined> {
```

Comment:
Can we split get and check token into 2 different methods. That way in generate we can still return the previous token if there was an issue generating a new one as it should still be valid for another 5 mins after we start trying to refresh it